### DISCUSSION TITLE
Visual Studio Code Remote - SSH not working

### DISCUSSION THREAD

**Discussion by laggingreflex** [termux/termux-packages#7439] [Q&A]
I'm trying to get [Visual Studio Code Remote - SSH][1] to work but I get this error:

> The remote host may not meet VS Code Server's prerequisites for glibc and libstdc++

<details> <summary>Screenshot</summary>

<a href="https://user-images.githubusercontent.com/2523987/131458843-2f542b2c-05e8-4fa5-9aee-0ecb246c0761.png"><img src="https://user-images.githubusercontent.com/2523987/131458843-2f542b2c-05e8-4fa5-9aee-0ecb246c0761.png" alt="screenshot"></img></a>

</details> 

<details>
<summary>Detailed logs</summary>
<pre>
[12:33:40.143] Log Level: 2
[12:33:40.144] remote-ssh@0.65.7
[12:33:40.145] win32 x64
[12:33:40.146] SSH Resolver called for "ssh-remote+xxx", attempt 1
[12:33:40.147] "remote.SSH.useLocalServer": false
[12:33:40.148] "remote.SSH.showLoginTerminal": false
[12:33:40.149] "remote.SSH.remotePlatform": {}
[12:33:40.149] "remote.SSH.path": undefined
[12:33:40.149] "remote.SSH.configFile": undefined
[12:33:40.150] "remote.SSH.useFlock": true
[12:33:40.150] "remote.SSH.lockfilesInTmp": false
[12:33:40.151] "remote.SSH.localServerDownload": auto
[12:33:40.151] "remote.SSH.remoteServerListenOnSocket": false
[12:33:40.152] "remote.SSH.showLoginTerminal": false
[12:33:40.152] "remote.SSH.defaultExtensions": []
[12:33:40.152] "remote.SSH.loglevel": 2
[12:33:40.153] SSH Resolver called for host: xxx
[12:33:40.153] Setting up SSH remote "xxx"
[12:33:40.163] Using commit id "3866c3553be8b268c8a7f8c0482c0c0177aa8bfa" and quality "stable" for server
[12:33:40.167] Install and start server if needed
[12:33:42.335] Checking ssh with "ssh -V"
[12:33:42.408] > OpenSSH_for_Windows_8.1p1, 
[12:33:42.408] > LibreSSL 3.0.2

[12:33:42.429] Running script with connection command: ssh -T -D 49184 xxx bash
[12:33:42.435] Terminal shell path: C:\Windows\System32\cmd.exe
[12:33:42.709] > ]0;C:\Windows\System32\cmd.exe
[12:33:42.709] Got some output, clearing connection timeout
[12:33:43.281] > b00c8153a624: running
[12:33:43.454] > Warning: Can't find libstdc++.so or ldconfig, can't verify libstdc++ version
[12:33:43.474] > bash: line 175: ldd: command not found
> Missing GLIBC >= 2.17!
> Found version
> b00c8153a624$$1$$
> Acquiring lock on /data/data/com.termux/files/home/.vscode-server/bin/3866c3553b
> e8b268c8a7f8c0482c0c0177aa8bfa/vscode-remote-lock.u0_a613.3866c3553be8b268c8a7f8
> c0482c0c0177aa8bfa
[12:33:43.514] > Found existing installation at /data/data/com.termux/files/home/.vscode-server/b
> in/3866c3553be8b268c8a7f8c0482c0c0177aa8bfa...
[12:33:43.520] > Checking /data/data/com.termux/files/home/.vscode-server/.3866c3553be8b268c8a7f8
> c0482c0c0177aa8bfa.log and /data/data/com.termux/files/home/.vscode-server/.3866
> c3553be8b268c8a7f8c0482c0c0177aa8bfa.pid for a running server
[12:33:43.536] > Looking for server with pid: 14389
[12:33:43.653] > Starting server with command... /data/data/com.termux/files/home/.vscode-server/
> bin/3866c3553be8b268c8a7f8c0482c0c0177aa8bfa/server.sh --start-server --host=127
> .0.0.1 --enable-remote-auto-shutdown --disable-telemetry --port=0  &> "/data/dat
> a/com.termux/files/home/.vscode-server/.3866c3553be8b268c8a7f8c0482c0c0177aa8bfa
> .log" < /dev/null
> printenv:
>     SHELL=/data/data/com.termux/files/usr/bin/bash
>     PREFIX=/data/data/com.termux/files/usr
>     PWD=/data/data/com.termux/files/home
>     LOGNAME=u0_a613
>     TERMUX_VERSION=0.116
>     EXTERNAL_STORAGE=/sdcard
>     LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
>     HOME=/data/data/com.termux/files/home
>     LANG=en_US.UTF-8
>     ANDROID_RUNTIME_ROOT=/apex/com.android.runtime
>     DEX2OATBOOTCLASSPATH=/apex/com.android.runtime/javalib/core-oj.jar:/apex/com
> .android.runtime/javalib/core-libart.jar:/apex/com.android.runtime/javalib/okhtt
> p.jar:/apex/com.android.runtime/javalib/bouncycastle.jar:/apex/com.android.runti
> me/javalib/apache-xml.jar:/system/framework/QPerformance.jar:/system/framework/U
> xPerformance.jar:/system/framework/framework.jar:/system/framework/ext.jar:/syst
> em/framework/telephony-common.jar:/system/framework/voip-common.jar:/system/fram
> ework/ims-common.jar:/system/framework/android.test.base.jar:/system/framework/t
> cmiface.jar:/system/framework/telephony-ext.jar:/system/framework/WfdCommon.jar:
> /system/framework/qcom.fmradio.jar:/system/framework/com.nxp.nfc.nq.jar
>     TMPDIR=/data/data/com.termux/files/usr/tmp
>     VSCODE_AGENT_FOLDER=/data/data/com.termux/files/home/.vscode-server
>     SSH_CONNECTION=192.168.0.100 49185 192.168.0.110 8022
>     ANDROID_DATA=/data
>     USER=u0_a613
>     SHLVL=0
>     ANDROID_ROOT=/system
>     BOOTCLASSPATH=/apex/com.android.runtime/javalib/core-oj.jar:/apex/com.androi
> d.runtime/javalib/core-libart.jar:/apex/com.android.runtime/javalib/okhttp.jar:/
> apex/com.android.runtime/javalib/bouncycastle.jar:/apex/com.android.runtime/java
> lib/apache-xml.jar:/system/framework/QPerformance.jar:/system/framework/UxPerfor
> mance.jar:/system/framework/framework.jar:/system/framework/ext.jar:/system/fram
> ework/telephony-common.jar:/system/framework/voip-common.jar:/system/framework/i
> ms-common.jar:/system/framework/android.test.base.jar:/system/framework/tcmiface
> .jar:/system/framework/telephony-ext.jar:/system/framework/WfdCommon.jar:/system
> /framework/qcom.fmradio.jar:/system/framework/com.nxp.nfc.nq.jar:/apex/com.andro
> id.conscrypt/javalib/conscrypt.jar:/apex/com.android.media/javalib/updatable-med
> ia.jar
>     SSH_CLIENT=192.168.0.100 49185 8022
>     ANDROID_TZDATA_ROOT=/apex/com.android.tzdata
>     PATH=/data/data/com.termux/files/usr/bin
>     _=/data/data/com.termux/files/usr/bin/printenv
[12:33:43.689] > Spawned remote server: 14988
[12:33:43.731] > Waiting for server log...
[12:33:51.649] > Waiting for server log...
[12:33:51.740] >  
> *
> * Reminder: You may only use this software with Visual Studio family products,
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057) 
> *
> 
[12:33:51.790] > Server did not start successfully. Full server log at /data/data/com.termux/file
> s/home/.vscode-server/.3866c3553be8b268c8a7f8c0482c0c0177aa8bfa.log >>>
[12:33:51.809] > /data/data/com.termux/files/home/.vscode-server/bin/3866c3553be8b268c8a7f8c0482c
> 0c0177aa8bfa/server.sh: 12: /data/data/com.termux/files/home/.vscode-server/bin/
> 3866c3553be8b268c8a7f8c0482c0c0177aa8bfa/node: not found
> <<< End of server log
> b00c8153a624: start
[12:33:51.814] > 
> exitCode==32==
> listeningOn====
> osReleaseId==Linux==
> arch==aarch64==
> tmpDir==/tmp==
> platform==linux==
> unpackResult====
> didLocalDownload==0==
[12:33:51.821] > 
> downloadTime====
> installTime====
> extInstallTime====
> serverStartTime==8101==
> b00c8153a624: end
[12:33:51.821] Received install output: 
exitCode==32==
listeningOn====
osReleaseId==Linux==
arch==aarch64==
tmpDir==/tmp==
platform==linux==
unpackResult====
didLocalDownload==0==
downloadTime====
installTime====
extInstallTime====
serverStartTime==8101==

[12:33:51.829] Resolver error: Error: The VS Code Server failed to start
	at Function.ServerInstallError (...\.vscode\extensions\ms-vscode-remote.remote-ssh-0.65.7\out\extension.js:1:64593)
	at c (...\.vscode\extensions\ms-vscode-remote.remote-ssh-0.65.7\out\extension.js:1:58918)
	at Object.t.handleInstallOutput (...\.vscode\extensions\ms-vscode-remote.remote-ssh-0.65.7\out\extension.js:1:63335)
	at Object.t.tryInstall (...\.vscode\extensions\ms-vscode-remote.remote-ssh-0.65.7\out\extension.js:1:415135)
	at processTicksAndRejections (internal/process/task_queues.js:93:5)
	at async ...\.vscode\extensions\ms-vscode-remote.remote-ssh-0.65.7\out\extension.js:1:294918
	at async Object.t.withShowDetailsEvent (...\.vscode\extensions\ms-vscode-remote.remote-ssh-0.65.7\out\extension.js:1:406463)
	at async Object.t.resolve (...\.vscode\extensions\ms-vscode-remote.remote-ssh-0.65.7\out\extension.js:1:295994)
	at async ...\.vscode\extensions\ms-vscode-remote.remote-ssh-0.65.7\out\extension.js:127:110656
[12:33:51.842] ------

[12:33:52.545] > 
[12:33:53.281] "install" terminal command done
[12:33:53.321] Install terminal quit with output: 

</pre>
</details>

I searched for similar issues but I'm not sure if they help me:

* https://github.com/termux/termux-packages/issues/121

  I'm not sure what this means:

  > You need to explicitly link with -lgnustl_shared for a more fully featured C++ library - see http://developer.android.com/ndk/guides/cpp-support.html#system

  Does it mean I have to compile Termux on my own?

  I tried to install gcc (which it suggested to install via package clang), but after installing running gcc gives this error:

  > CANNOT LINK EXECUTABLE "gcc": cannot locate symbol "ffi_type_double" referenced by "/data/data/com.termux/files/usr/lib/libLLVM-12.so"...

* https://github.com/termux/termux-packages/issues/6316
* https://github.com/termux/termux-packages/issues/5581

  Not sure if these have to do with VSCode Remote SSH?

  I've tried to install code-server but it gave error too:

  > CANNOT LINK EXECUTABLE "cc": cannot locate symbol "ffi_type_double" referenced by "/data/data/com.termux/files/usr/lib/libLLVM-12.so"...

  <details> <summary> Detailed logs </summary> <pre>
    $ npm i code-server

    > argon2@0.28.2 install /data/data/com.termux/files/home/node_modules/argon2
    > node-pre-gyp install --fallback-to-build

    node-pre-gyp ERR! install response status 404 Not Found on https://github.com/ranisalt/node-argon2/releases/download/v0.28.2/argon2-v0.28.2-napi-v3-android-arm64-unknown.tar.gz
    node-pre-gyp WARN Pre-built binaries not installable for argon2@0.28.2 and node@14.15.4 (node-v83 ABI, unknown) (falling back to source compile with node-gyp)
    node-pre-gyp WARN Hit error response status 404 Not Found on https://github.com/ranisalt/node-argon2/releases/download/v0.28.2/argon2-v0.28.2-napi-v3-android-arm64-unknown.tar.gz
    make: Entering directory '/data/data/com.termux/files/home/node_modules/argon2/build'
      CC(target) Release/obj.target/libargon2/argon2/src/ref.o
    CANNOT LINK EXECUTABLE "cc": cannot locate symbol "ffi_type_double" referenced by "/data/data/com.termux/files/usr/lib/libLLVM-12.so"...
    make: *** [libargon2.target.mk:117: Release/obj.target/libargon2/argon2/src/ref.o] Error 1
    make: Leaving directory '/data/data/com.termux/files/home/node_modules/argon2/build'
    gyp ERR! build error
    gyp ERR! stack Error: `make` failed with exit code: 2
    gyp ERR! stack     at ChildProcess.onExit (/data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
    gyp ERR! stack     at ChildProcess.emit (events.js:315:20)
    gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)
    gyp ERR! System Linux 4.4.192-perf+
    gyp ERR! command "/data/data/com.termux/files/usr/bin/node" "/data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "build" "--fallback-to-build" "--module=/data/data/com.termux/files/home/node_modules/argon2/lib/binding/napi-v3/argon2.node" "--module_name=argon2" "--module_path=/data/data/com.termux/files/home/node_modules/argon2/lib/binding/napi-v3" "--napi_version=7" "--node_abi_napi=napi" "--napi_build_version=3" "--node_napi_label=napi-v3"
    gyp ERR! cwd /data/data/com.termux/files/home/node_modules/argon2
    gyp ERR! node -v v14.15.4
    gyp ERR! node-gyp -v v5.1.0
    gyp ERR! not ok
    node-pre-gyp ERR! build error
    node-pre-gyp ERR! stack Error: Failed to execute '/data/data/com.termux/files/usr/bin/node /data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/data/data/com.termux/files/home/node_modules/argon2/lib/binding/napi-v3/argon2.node --module_name=argon2 --module_path=/data/data/com.termux/files/home/node_modules/argon2/lib/binding/napi-v3 --napi_version=7 --node_abi_napi=napi --napi_build_version=3 --node_napi_label=napi-v3' (1)
    node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/data/data/com.termux/files/home/node_modules/@mapbox/node-pre-gyp/lib/util/compile.js:89:23)
    node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:315:20)
    node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:1048:16)
    node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)
    node-pre-gyp ERR! System Linux 4.4.192-perf+
    node-pre-gyp ERR! command "/data/data/com.termux/files/usr/bin/node" "/data/data/com.termux/files/home/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
    node-pre-gyp ERR! cwd /data/data/com.termux/files/home/node_modules/argon2
    node-pre-gyp ERR! node -v v14.15.4
    node-pre-gyp ERR! node-pre-gyp -v v1.0.5
    node-pre-gyp ERR! not ok
    Failed to execute '/data/data/com.termux/files/usr/bin/node /data/data/com.termux/files/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/data/data/com.termux/files/home/node_modules/argon2/lib/binding/napi-v3/argon2.node --module_name=argon2 --module_path=/data/data/com.termux/files/home/node_modules/argon2/lib/binding/napi-v3 --napi_version=7 --node_abi_napi=napi --napi_build_version=3 --node_napi_label=napi-v3' (1)
    npm WARN enoent ENOENT: no such file or directory, open '/data/data/com.termux/files/home/package.json'
    npm WARN @coder/logger@1.1.16 requires a peer of @google-cloud/logging@^4.5.2 but none is installed. You must install peer dependencies yourself.
    npm WARN home No description
    npm WARN home No repository field.
    npm WARN home No README data
    npm WARN home No license field.

    npm ERR! code ELIFECYCLE
    npm ERR! errno 1
    npm ERR! argon2@0.28.2 install: `node-pre-gyp install --fallback-to-build`
    npm ERR! Exit status 1
    npm ERR!
    npm ERR! Failed at the argon2@0.28.2 install script.
    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

    npm ERR! A complete log of this run can be found in:
    npm ERR!     /data/data/com.termux/files/home/.npm/_logs/2021-08-31T07_43_26_042Z-debug.log  
  </pre> </details>

I'm not sure what else to do. 

Any ideas how to get VS Code Remote - SSH to work on Termux on Android?

I'm using: 

* Phone: Android 10, Termux v0.116
* Desktop: Windows 10, VS Code v1.59.1

[1]: https://code.visualstudio.com/docs/remote/ssh
[2]: https://user-images.githubusercontent.com/2523987/131458843-2f542b2c-05e8-4fa5-9aee-0ecb246c0761.png

**Comment by MrAdityaAlok** [comment]
`glibc` != `bionic libc` which Android and hence Termux uses. This should  be supported  by Microsoft to work.

From where did you get `gcc` ? If `its-pointless` then it is known to be some what buggy. Termux official repo only provides `clang` and `gcc` is symlink to it.

And code-server works well with Termux . You can see here : https://github.com/cdr/code-server/blob/main/docs/termux.md

**Reply by laggingreflex** [reply]
> From where did you get gcc ?

Via clang

I think the main issue is this error:

> Missing GLIBC >= 2.17!

I tried to [check glib version][1] via two recommended methods, neither of which seem to work:

* By running `$ ldd --version`

  It gives error:

  ```sh
  $ ldd --version    
  objdump: '--version': No such file
  ```
  I checked the same command in on an Ubuntu machine where it works correctly: 
  ```sh
  $ ldd --version
  ldd (Ubuntu GLIBC 2.31-0ubuntu9) 2.31
  ```
  So the problem is probably Android/Termux. 


* By writing a test program:

  ```c
  #include <stdio.h>
  #include <stdlib.h>
  #include <gnu/libc-version.h>

  int main(int argc, char *argv[]) {
    printf("GNU libc version: %s\n", gnu_get_libc_version());
    exit(EXIT_SUCCESS);
  }  
  ```
  But compiling it gives error:

  ```sh
  $ gcc glibc-version.c -o glibc-version
  glibc-version.c:3:10: fatal error: 'gnu/libc-version.h' file not found
  #include <gnu/libc-version.h>
          ^~~~~~~~~~~~~~~~~~~~
  1 error generated.
  ```

Is there any other way to solve this: 

> Missing GLIBC >= 2.17!

[1]: https://stackoverflow.com/questions/9705660/check-glibc-version-for-a-particular-gcc-compiler

**Reply by laggingreflex** [reply]
> And code-server works well with Termux

Is code-server same as [VS Code Remote - SSH][1]? 

I've manage to install code-server but I'm trying to get the latter to work. 

[1]: https://code.visualstudio.com/docs/remote/ssh

**Comment by MrAdityaAlok** [comment]
> Is code-server same as VS Code Remote - SSH?

No, it is vscode in browser. You can run code server in termux and access it through any browser on any device in your local network (or externally by port forwarding with services something like [ngrok](https://ngrok.com))

Related to glibc:
- As stated previously Android uses it's own implementation of standard c ([bionic](https://en.m.wikipedia.org/wiki/Bionic_(software))).  So you cannot uses gnu headers directly. It is based on code from OpenBSD, rather than glibc.
- And also `gcc` is not the gnu c compiler it is just symlink to clang.
```
$ gcc --version
clang version 12.0.1
Target: aarch64-unknown-linux-android24
Thread model: posix
InstalledDir: /data/data/com.termux/files/usr/bin
```

If still you want to use `glibc` you should use ubuntu or any other distribution in proot in termux. More info [here](https://github.com/termux/proot-distro)
